E0308 doesn't provide a help message in loops when using `let-else`

### Code

```Rust
pub fn main() {
    let a = loop {
        if true {
            break;
        }
        let Some(_) = Some(5) else {
            break 3;
        };
    };
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:7:19
  |
7 |             break 3;
  |                   ^ expected `()`, found integer

For more information about this error, try `rustc --explain E0308`.
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:9:19
  |
4 |             break;
  |             ----- expected because of this `break`
...
7 |             break 3;
  |                   ^ expected `()`, found integer

For more information about this error, try `rustc --explain E0308`.
```

### Rationale and extra context

Usually, the error comes with a nice help message showing _why_ another type was expected. But when using `let-else`, no help message is shown.

PS: in all repros, replacing `if true { break; }` with just `break;` works just as well -- but it also causes an `unreachable_code` warning, which I felt like muddied the waters.

### Other cases

```Rust
// the following 3 changes bring the help message back:

// replacing `let-else` with logically identical `if-let`
pub fn main() {
    let a = loop {
        if true {
            break;
        }
        if let Some(_n) = Some(5) {
            //
        } else {
            break 3;
        };
    };
}

// replacing `let-else` with `if`
pub fn main() {
    let a = loop {
        if true {
            break;
        }
        if true {
            break 3;
        }
    };
}

// having `let-else` contain the _first_ `break` of the loop
pub fn main() {
    let a = loop {
        let Some(_) = Some(5) else {
            break;
        };
        if true {
            break 3;
        }
    };
}

// having two `let-else`s doesn't work though
pub fn main() {
    let a = loop {
        let Some(_) = Some(5) else {
            break;
        };
        let Some(_) = Some(4) else {
            break 3;
        };
    };
}
```

### Rust Version

```Shell
rustc 1.86.0 (05f9846f8 2025-03-31) (built from a source tarball)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: x86_64-unknown-linux-gnu
release: 1.86.0
LLVM version: 19.1.7
```

### Anything else?

All the repros were tested on Rust Playground

--- a/compiler/rustc_hir_typeck/src/demand.rs
+++ b/compiler/rustc_hir_typeck/src/demand.rs
@@ -605,6 +605,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         parent_id = self.tcx.parent_hir_id(*hir_id);
                         parent
                     }
+                    hir::Node::Stmt(hir::Stmt { hir_id, kind: hir::StmtKind::Let(_), .. }) => {
+                        parent_id = self.tcx.parent_hir_id(*hir_id);
+                        parent
+                    }
+                    hir::Node::LetStmt(hir::LetStmt { hir_id, .. }) => {
+                        parent_id = self.tcx.parent_hir_id(*hir_id);
+                        parent
+                    }
                     hir::Node::Block(_) => {
                         parent_id = self.tcx.parent_hir_id(parent_id);
                         parent

--- a/tests/ui/let-else/let-else-break-help-issue-142602.rs
+++ b/tests/ui/let-else/let-else-break-help-issue-142602.rs
@@ -1,0 +1,23 @@
+// testcase from https://github.com/rust-lang/rust/issues/142602
+
+pub fn main() {
+    // Case 1: break before let-else
+    let _a = loop {
+        if true {
+            break;
+        }
+        let Some(_) = Some(5) else {
+            break 3; //~ ERROR mismatched types
+        };
+    };
+
+    // Case 2: two let-else statements
+    let _b = loop {
+        let Some(_) = Some(5) else {
+            break;
+        };
+        let Some(_) = Some(4) else {
+            break 3; //~ ERROR mismatched types
+        };
+    };
+}

--- a/tests/ui/let-else/let-else-break-help-issue-142602.stderr
+++ b/tests/ui/let-else/let-else-break-help-issue-142602.stderr
@@ -1,0 +1,21 @@
+error[E0308]: mismatched types
+  --> $DIR/let-else-break-help-issue-142602.rs:10:19
+   |
+LL |             break;
+   |             ----- expected because of this `break`
+...
+LL |             break 3;
+   |                   ^ expected `()`, found integer
+
+error[E0308]: mismatched types
+  --> $DIR/let-else-break-help-issue-142602.rs:20:19
+   |
+LL |             break;
+   |             ----- expected because of this `break`
+...
+LL |             break 3;
+   |                   ^ expected `()`, found integer
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

